### PR TITLE
fix float types in Phi (use float16)

### PR DESCRIPTION
**Pull request by davidkoski**
- per suggestions in #23 ensure that the values that go into the cache are float16

### CODE REVIEW

File: Libraries/LLM/Phi.swift [hunk 1]
```diff
@@ -73,14 +73,12 @@ private class PhiAttention: Module {
...
-
         // Finally perform the attention computation
         let scale = sqrt(1 / Float(queries.dim(-1)))
         let output = MLXFast.scaledDotProductAttention(
-            queries: queries, keys: keys, values: values, scale: scale, mask: mask
+            queries: queries.asType(.float32), keys: keys, values: values, scale: scale, mask: mask
         )
+        .asType(values.dtype)
         .transposed(0, 2, 1, 3)
         .reshaped(B, L, -1)
```
Comment (davidkoski) [inline Libraries/LLM/Phi.swift:83]: This provides: 

```
output float16, keys float16, values float16
```

File: Libraries/LLM/Phi.swift [hunk 2]
```diff
@@ -73,14 +73,12 @@ private class PhiAttention: Module {
...
         }
 
-        queries = queries.asType(Float.self)
-        keys = keys.asType(Float.self)
-
         // Finally perform the attention computation
         let scale = sqrt(1 / Float(queries.dim(-1)))
         let output = MLXFast.scaledDotProductAttention(
-            queries: queries, keys: keys, values: values, scale: scale, mask: mask
+            queries: queries.asType(.float32), keys: keys, values: values, scale: scale, mask: mask
```
Comment (davidkoski) [inline Libraries/LLM/Phi.swift:79]: I found that if I passed `queries` in as Float16 I would get unimaginative output

image

Is this a reasonable thing to do here?  Should keys be cast to float32 as well (it didn't seem to matter when I tested it)?

@awni

**Review by awni — APPROVED** [review]
🚀

File: Libraries/LLM/Phi.swift (hunk 2 above)
Comment (awni) [reply Libraries/LLM/Phi.swift:79]: So if you upcast the queries the full attention should be done in fp32 and it should work. The keys will be implicitly cast in the op.

The way you’ve done it looks good to me. Most models don’t need this, phi2 is an exception for some reason

**Pull request merged**